clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -1049,236 +1050,362 @@ CodeGenFunction::evaluateOrEmitBuiltinObjectSize(const Expr *E, unsigned Type,
+  const ArrayType *ArrayTy = Ctx.getAsArrayType(FlexibleArrayMemberTy);
+  CharUnits BaseSize = Ctx.getTypeSizeInChars(ArrayTy->getElementType());
+  auto *FlexibleArrayMemberBaseSize =
+      llvm::ConstantInt::get(ResType, BaseSize.getQuantity(), IsSigned);
+
+  //  size_t flexible_array_member_size =
+  //          count * flexible_array_member_base_size;
+  Value *FlexibleArrayMemberSize =
+      Builder.CreateMul(Count, FlexibleArrayMemberBaseSize,
+                        "flexible_array_member_size", !IsSigned, IsSigned);
+
+  Value *Res = nullptr;
+  if (FlexibleArrayMemberFD == FD) {
+    if (Idx) { // Option (2) '&ptr->array[idx]'
+      //  size_t index_size = index * flexible_array_member_base_size;
+      Value *IndexSize = Builder.CreateMul(FlexibleArrayMemberBaseSize, Index,
+                                           "index_size", !IsSigned, IsSigned);
+
+      //  return flexible_array_member_size - index_size;
+      Res = Builder.CreateSub(FlexibleArrayMemberSize, IndexSize, "result",
+                              !IsSigned, IsSigned);
+    } else { // Option (1) 'ptr->array'
+      //  return flexible_array_member_size;
+      Res = FlexibleArrayMemberSize;
+    }
+  } else {
+    //  size_t sizeof_struct = sizeof (struct s);
+    llvm::StructType *StructTy = getTypes().getCGRecordLayout(RD).getLLVMType();
+    const llvm::DataLayout &Layout = CGM.getDataLayout();
+    TypeSize Size = Layout.getTypeSizeInBits(StructTy);
+    Value *SizeofStruct =
+        llvm::ConstantInt::get(ResType, Size.getKnownMinValue() / CharWidth);
+
+    if (Idx) { // Option (4) '&ptr->field_array[idx]'
+      //  size_t field_base_size = sizeof (*ptr->field_array);
+      const ArrayType *ArrayTy = Ctx.getAsArrayType(FieldTy);
+      CharUnits BaseSize = Ctx.getTypeSizeInChars(ArrayTy->getElementType());
+      auto *FieldBaseSize =
+          llvm::ConstantInt::get(ResType, BaseSize.getQuantity(), IsSigned);
+
+      //  field_offset += index * field_base_size;
+      Value *Mul = Builder.CreateMul(Index, FieldBaseSize, "field_offset",
+                                     !IsSigned, IsSigned);
+      FieldOffset = Builder.CreateAdd(FieldOffset, Mul);
+
+      //  size_t offset_diff = flexible_array_member_offset - field_offset;
+      Value *OffsetDiff = Builder.CreateSub(SizeofStruct, FieldOffset,
+                                            "offset_diff", !IsSigned, IsSigned);
+
+      //  return offset_diff + flexible_array_member_size;
+      Res = Builder.CreateAdd(FlexibleArrayMemberSize, OffsetDiff, "result");
+    } else { // Option (3) '&ptr->field'
+      //  size_t offset_diff = flexible_array_member_offset - field_offset;
+      Value *OffsetDiff = Builder.CreateSub(SizeofStruct, FieldOffset,
+                                            "offset_diff", !IsSigned, IsSigned);
+
+      //  return flexible_array_member_size + offset_diff;
+      Res = Builder.CreateAdd(FlexibleArrayMemberSize, OffsetDiff, "result");
+    }
```

Comment:
You can now tail-merge these two cases.

```c++
if (Idx) {
...
} // else is no longer necessary (Option 3)
//  size_t offset_diff = flexible_array_member_offset - field_offset;
Value *OffsetDiff = Builder.CreateSub(SizeofStruct, FieldOffset,
                                            "offset_diff", !IsSigned, IsSigned);

//  return flexible_array_member_size + offset_diff;
Res = Builder.CreateAdd(FlexibleArrayMemberSize, OffsetDiff, "result");
```

Reply:
I'm not able to edit things locally on my computer, because GitHub fucked up. So I won't be able to test any code changes I make. Let's leave this be.